services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -154,9 +91,10 @@ private enum Action {
     private final RealmModel realm;
     private final EventBuilder event;
 
-    private Action action;
-
     private String grantType;
+    private List<OAuth2GrantType> grants;
+    private OAuth2GrantType grant;
+    private OAuth2GrantType.Context context;
```

Comment:
This is not yet fully changed, but I don't consider this as a blocker of this PR (`context` can be likely declared as "method scoped"  variable instead of the "object scoped" variable. The `grant` can be a bit more tricky as it is currently used in more methods, so it is completely fine to keep it as it is IMO).